test/hotspot/jtreg/gc/TestTransparentHugePagesHeap.java

```diff
@@ -0,0 +1,152 @@
+*/
+/*
+ * @test id=Parallel
+ * @summary Run tests with Parallel
+ * @library /test/lib
+ * @build jdk.test.lib.Platform
+ * @requires os.family == "linux"
+ * @requires vm.gc.Parallel
+ * @run driver TestTransparentHugePagesHeap Parallel
+*/
+/*
+ * @test id=Serial
+ * @summary Run tests with Serial
+ * @library /test/lib
+ * @build jdk.test.lib.Platform
+ * @requires os.family == "linux"
+ * @requires vm.gc.Serial
+ * @run driver TestTransparentHugePagesHeap Serial
+*/
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.Scanner;
+
+import jdk.test.lib.os.linux.HugePageConfiguration;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.Platform;
+
+import jtreg.SkippedException;
+
+// We verify that the heap can be backed by THP by looking at the
+// THPeligible field for the heap section in /proc/self/smaps. This
+// field indicates if a mapping can use THP.
+// THP mode 'always': this field is 1 whenever huge pages can be used
+// THP mode 'madvise': this field is 1 if the mapping has been madvised
+// as MADV_HUGEPAGE. In the JVM that should happen when the flag
+// -XX:+UseTransparentHugePages is specified.
+//
+// Note: we don't verify if the heap is backed by huge pages because we
+// can't know if the underlying system have any available.
+public class TestTransparentHugePagesHeap {
+
+    public static void main(String args[]) throws Exception {
+        // To be able to detect large page use (esp. THP) somewhat reliably, we
+        //  need at least kernel 3.8 to get the "VmFlags" tag in smaps.
+        // (Note: its still good we started the VM at least since this serves as a nice
+        //  test for all manners of large page options).
+        if (Platform.getOsVersionMajor() < 3 ||
+            (Platform.getOsVersionMajor() == 3 && Platform.getOsVersionMinor() < 8)) {
+            throw new SkippedException("Kernel older than 3.8 - skipping this test.");
+        }
+
+        final HugePageConfiguration hugePageConfiguration = HugePageConfiguration.readFromOS();
+        if (!hugePageConfiguration.supportsTHP()) {
+            throw new SkippedException("THP is turned off");
+        }
+
+        OutputAnalyzer oa = ProcessTools.executeTestJava("-XX:+Use" + args[0] + "GC", "-Xmx128m", "-Xms128m", "-Xlog:pagesize:thp-%p.log", "-XX:+UseTransparentHugePages", VerifyTHPEnabledForHeap.class.getName());
+        oa.shouldHaveExitValue(0);
+    }
+
+    class VerifyTHPEnabledForHeap {
+
+        public static void main(String args[]) throws Exception {
+            String heapAddress = readHeapAddressInLog();
+            Path smaps = makeSmapsCopy();
+
+            final Pattern heapSection = Pattern.compile("^" + heapAddress + ".*");
+            final Pattern thpEligible = Pattern.compile("THPeligible:\\s+(\\d)\\s*");
```

Comment:
I thought we had the `smaps` parser somewhere, and here it is: https://github.com/openjdk/jdk/blob/444a8fa14e8ab016b8aae018054c5dc1eb843fee/test/hotspot/jtreg/runtime/os/TestTracePageSizes.java#L149 -- maybe yank it from there to some testlib location, and use it?

Reply:
I agree that this is what we want long-term, we talked about this internally as well. At one point we also just extended the `TestTracePageSizes.java` test to ensure THP for the heap to avoid some duplication. In the end we felt creating a new test was the best approach right now, to also allow for small and safe backport to 25u.

Created [JDK-8366716](https://bugs.openjdk.org/browse/JDK-8366716) to keep track of this. 

Are you good with this approach?